File: object_relations/app/models/restaurant.rb
class Restaurant

  @@all = []

  attr_reader :name

  def initialize(name)
    @name = name
    @@all << self
  end

  def reviews
    Review.all.select {|review| review.restaurant==self}
  end

  def customers
    reviews.map {|review| review.customer}.uniq
  end

  def self.all
    @@all
  end

  def self.find_by_name(name)
    all.find {|restaurant| restaurant.name==name}
  end

end

File: javascript/models/restaurant.js
const Restaurant = ( function Restaurant(){
  var counter = 0

  return class {
    constructor(name) {
      this.id = ++counter
      this.name = name
      store.restaurants.push(this)
    }

    reviews(){
      return store.reviews.filter((review) => {
        return review.restaurantId===this.id
      })
    }

    customers(){
      return this.reviews().map((review) => {
        return review.customer()[0]
      })
    }

    static all() {
      return store.restaurants
    }
  }
}
())
